Invalid conditional.type.incompatible error

Consider the following code:
`import java.io.IOException;
import java.io.Reader;

import org.checkerframework.checker.index.qual.*;

class conderr extends Reader{
public Reader abc;
public conderr(Reader xyz){
	this.abc = xyz;
}
public void close() throws IOException {
	abc.close();
}
public int read( final char[] cbuf, final @NonNegative int off, final int len ) throws IOException { // offset value has to be non negative
        int c;
        for (@NonNegative int i = 0; i < len; i++ ) {
            c = super.read();
            if ( c == -1 ) {
                return i == 0 ? -1 : i;
            }
        }
        return len;
    }
}`
Type in the terminal: javac -processor org.checkerframework.checker.index.IndexChecker conderr.java
Output:
conderr.java:14: error: [override.return.invalid] Incompatible return type.
public int read( final char[] cbuf, final @NonNegative int off, final int len ) throws IOException { // offset value has to be non negative
       ^
  Method
    @LowerBoundUnknown int read(@LowerBoundUnknown conderr this, @LowerBoundUnknown char @LowerBoundUnknown [] p0, @NonNegative int p1, @LowerBoundUnknown int p2) throws @LowerBoundUnknown IOException in conderr
  cannot override
    @GTENegativeOne int read(@LowerBoundUnknown Reader this, @LowerBoundUnknown char @LowerBoundUnknown [] p0, @NonNegative int p1, @NonNegative int p2) throws @LowerBoundUnknown IOException in java.io.Reader
  found   : @LowerBoundUnknown int
  required: @GTENegativeOne int
conderr.java:14: error: [override.return.invalid] Incompatible return type.
public int read( final char[] cbuf, final @NonNegative int off, final int len ) throws IOException { // offset value has to be non negative
       ^
  Method
    @UpperBoundUnknown int read(@UpperBoundUnknown conderr this, @UpperBoundUnknown char @UpperBoundUnknown [] p0, @UpperBoundUnknown int p1, @UpperBoundUnknown int p2) throws @UpperBoundUnknown IOException in conderr
  cannot override
    @LTEqLengthOf("#1") int read(@UpperBoundUnknown Reader this, @UpperBoundUnknown char @UpperBoundUnknown [] p0, @LTEqLengthOf("#1") int p1, @LTEqLengthOf("#1") int p2) throws @UpperBoundUnknown IOException in java.io.Reader
  found   : @UpperBoundUnknown int
  required: @LTEqLengthOf("#1") int
**conderr.java:19: error: [conditional.type.incompatible] incompatible types in conditional expression.
                return i == 0 ? -1 : i;
                                     ^
  found   : @IntRangeFromNonNegative int
  required: @IntVal(-1) int**
3 errors

Notice the last error. There is no need for i to be annotated with @IntVal(-1), it can take any non negative integral value. I think it is a bug.
@kelloggm

--- a/checker/tests/index/IndexConditionalReport.java
+++ b/checker/tests/index/IndexConditionalReport.java
@@ -1,0 +1,13 @@
+// test case for https://github.com/typetools/checker-framework/issues/2345
+
+class IndexConditionalReport {
+
+    public int getI(int len) {
+        for (int i = 0; i < len; i++) {
+            if (false) {
+                return i == 0 ? -1 : i; // unexpected error issued here
+            }
+        }
+        return -1;
+    }
+}

--- a/framework/src/main/java/org/checkerframework/common/value/ValueAnnotatedTypeFactory.java
+++ b/framework/src/main/java/org/checkerframework/common/value/ValueAnnotatedTypeFactory.java
@@ -2223,15 +2223,19 @@ public class ValueAnnotatedTypeFactory extends BaseAnnotatedTypeFactory {
 
     @Override
     public AnnotatedTypeMirror getAnnotatedType(Tree tree) {
-        if (tree instanceof ConditionalExpressionTree) {
+        AnnotatedTypeMirror atm = super.getAnnotatedType(tree);
+
+        if (atm.isAnnotatedInHierarchy(UNKNOWNVAL) && tree instanceof ConditionalExpressionTree) {
             AnnotatedTypeMirror atmTrue =
                     getAnnotatedType(((ConditionalExpressionTree) tree).getTrueExpression());
             AnnotatedTypeMirror atmFalse =
                     getAnnotatedType(((ConditionalExpressionTree) tree).getFalseExpression());
             TypeMirror alub = TreeUtils.typeOf(tree);
-            return AnnotatedTypes.leastUpperBound(this, atmTrue, atmFalse, alub);
+            atm.replaceAnnotation(
+                    AnnotatedTypes.leastUpperBound(this, atmTrue, atmFalse, alub)
+                            .getEffectiveAnnotationInHierarchy(UNKNOWNVAL));
         }
 
-        return super.getAnnotatedType(tree);
+        return atm;
     }
 }

--- a/framework/src/main/java/org/checkerframework/common/value/ValueAnnotatedTypeFactory.java
+++ b/framework/src/main/java/org/checkerframework/common/value/ValueAnnotatedTypeFactory.java
@@ -67,7 +67,6 @@ import org.checkerframework.framework.type.treeannotator.PropagationTreeAnnotato
 import org.checkerframework.framework.type.treeannotator.TreeAnnotator;
 import org.checkerframework.framework.type.typeannotator.ListTypeAnnotator;
 import org.checkerframework.framework.type.typeannotator.TypeAnnotator;
-import org.checkerframework.framework.util.AnnotatedTypes;
 import org.checkerframework.framework.util.FieldInvariants;
 import org.checkerframework.framework.util.FlowExpressionParseUtil.FlowExpressionParseException;
 import org.checkerframework.framework.util.MultiGraphQualifierHierarchy;
@@ -1452,6 +1451,14 @@ public class ValueAnnotatedTypeFactory extends BaseAnnotatedTypeFactory {
             */
             return COVERED_CLASS_STRINGS.contains(tm.toString());
         }
+
+        @Override
+        public Void visitConditionalExpression(
+                ConditionalExpressionTree node, AnnotatedTypeMirror annotatedTypeMirror) {
+            // Work around for https://github.com/typetools/checker-framework/issues/602.
+            annotatedTypeMirror.replaceAnnotation(UNKNOWNVAL);
+            return super.visitConditionalExpression(node, annotatedTypeMirror);
+        }
     }
 
     /**
@@ -2219,23 +2226,5 @@ public class ValueAnnotatedTypeFactory extends BaseAnnotatedTypeFactory {
         }
 
         return getMinLenValue(lengthAnno);
-    }
-
-    @Override
-    public AnnotatedTypeMirror getAnnotatedType(Tree tree) {
-        AnnotatedTypeMirror atm = super.getAnnotatedType(tree);
-
-        if (atm.isAnnotatedInHierarchy(UNKNOWNVAL) && tree instanceof ConditionalExpressionTree) {
-            AnnotatedTypeMirror atmTrue =
-                    getAnnotatedType(((ConditionalExpressionTree) tree).getTrueExpression());
-            AnnotatedTypeMirror atmFalse =
-                    getAnnotatedType(((ConditionalExpressionTree) tree).getFalseExpression());
-            TypeMirror alub = TreeUtils.typeOf(tree);
-            atm.replaceAnnotation(
-                    AnnotatedTypes.leastUpperBound(this, atmTrue, atmFalse, alub)
-                            .getEffectiveAnnotationInHierarchy(UNKNOWNVAL));
-        }
-
-        return atm;
     }
 }

--- a/framework/src/main/java/org/checkerframework/common/value/ValueAnnotatedTypeFactory.java
+++ b/framework/src/main/java/org/checkerframework/common/value/ValueAnnotatedTypeFactory.java
@@ -1457,7 +1457,7 @@ public class ValueAnnotatedTypeFactory extends BaseAnnotatedTypeFactory {
                 ConditionalExpressionTree node, AnnotatedTypeMirror annotatedTypeMirror) {
             // Work around for https://github.com/typetools/checker-framework/issues/602.
             annotatedTypeMirror.replaceAnnotation(UNKNOWNVAL);
-            return super.visitConditionalExpression(node, annotatedTypeMirror);
+            return null;
         }
     }
 

--- a/framework/src/main/java/org/checkerframework/common/value/ValueAnnotatedTypeFactory.java
+++ b/framework/src/main/java/org/checkerframework/common/value/ValueAnnotatedTypeFactory.java
@@ -1,5 +1,6 @@
 package org.checkerframework.common.value;
 
+import com.sun.source.tree.ConditionalExpressionTree;
 import com.sun.source.tree.ExpressionTree;
 import com.sun.source.tree.LiteralTree;
 import com.sun.source.tree.MemberSelectTree;
@@ -66,6 +67,7 @@ import org.checkerframework.framework.type.treeannotator.PropagationTreeAnnotato
 import org.checkerframework.framework.type.treeannotator.TreeAnnotator;
 import org.checkerframework.framework.type.typeannotator.ListTypeAnnotator;
 import org.checkerframework.framework.type.typeannotator.TypeAnnotator;
+import org.checkerframework.framework.util.AnnotatedTypes;
 import org.checkerframework.framework.util.FieldInvariants;
 import org.checkerframework.framework.util.FlowExpressionParseUtil.FlowExpressionParseException;
 import org.checkerframework.framework.util.MultiGraphQualifierHierarchy;
@@ -2217,5 +2219,19 @@ public class ValueAnnotatedTypeFactory extends BaseAnnotatedTypeFactory {
         }
 
         return getMinLenValue(lengthAnno);
+    }
+
+    @Override
+    public AnnotatedTypeMirror getAnnotatedType(Tree tree) {
+        if (tree instanceof ConditionalExpressionTree) {
+            AnnotatedTypeMirror atmTrue =
+                    getAnnotatedType(((ConditionalExpressionTree) tree).getTrueExpression());
+            AnnotatedTypeMirror atmFalse =
+                    getAnnotatedType(((ConditionalExpressionTree) tree).getFalseExpression());
+            TypeMirror alub = TreeUtils.typeOf(tree);
+            return AnnotatedTypes.leastUpperBound(this, atmTrue, atmFalse, alub);
+        }
+
+        return super.getAnnotatedType(tree);
     }
 }

--- a/framework/tests/value/Repo.java
+++ b/framework/tests/value/Repo.java
@@ -9,8 +9,6 @@ class Repo {
         for (int i = 0; i < 20; i++) {
             // :: error: (assignment.type.incompatible)
             @IntVal(0) int x = i;
-            // TODO: this is a dataflow bug. (March, 6, 2015)
-            // :: error: (conditional.type.incompatible)
             int j = flag ? i : 3;
         }
     }

